src/back-end/lib/db.ts

```diff
@@ -654,3 +656,434 @@ export async function hasFilePermission(connection: Connection, session: Session
+    .where({ 'opp.id': id })
+    // Join on latest CWU status
+    .join<RawCWUOpportunity>('cwuOpportunityStatuses as stat', function() {
+      this
+        .on('opp.id', '=', 'stat.opportunity')
+        .andOn('stat.createdAt', '=',
+          connection.raw('(select max("createdAt") from "cwuOpportunityStatuses" as stat2 where \
+            stat2.opportunity = opp.id)'));
+    })
+    // Join on latest CWU version
+    .join<RawCWUOpportunity>('cwuOpportunityVersions as version', function() {
+      this
+        .on('opp.id', '=', 'version.opportunity')
+        .andOn('version.createdAt', '=',
+          connection.raw('(select max("createdAt") from "cwuOpportunityVersions" as version2 where \
+            version2.opportunity = opp.id)'));
+    })
+    .select<RawCWUOpportunity>([
+      'opp.id',
+      'opp.createdAt',
+      'opp.createdBy',
+      'version.createdAt as updatedAt',
+      'version.createdBy as updatedBy',
+      'version.title',
+      'version.teaser',
+      'version.remoteOk',
+      'version.remoteDesc',
+      'version.location',
+      'version.reward',
+      'version.skills',
+      'version.description',
+      'version.proposalDeadline',
+      'version.assignmentDate',
+      'version.startDate',
+      'version.completionDate',
+      'version.submissionInfo',
+      'version.acceptanceCriteria',
+      'version.evaluationCriteria',
+      'stat.status'
+    ]);
+
+  if (!session.user || session.user.type === UserType.Vendor) {
+    // Anonymous users and vendors can only see public opportunities
+    query = query
+      .whereIn('stat.status', publicOpportunityStatuses);
+  } else if (session.user.type === UserType.Government) {
+    // Gov users should only see private opportunities they own, and public opportunities
+    query = query
+      .whereIn('stat.status', publicOpportunityStatuses)
+      .orWhere(function() {
+        this
+          .whereIn('stat.status', privateOpportunitiesStatuses)
+          .andWhere({ 'opp.createdBy': session.user?.id });
+      });
+  } else {
+    // Admin users can see both private and public opportunities
+    query = query
+      .whereIn('stat.status', [...publicOpportunityStatuses, ...privateOpportunitiesStatuses]);
+  }
+
+  const result = await query.first();
+
+  // Query for attachment file ids
+  if (result) {
+    result.attachments = (await connection<{ id: Id }>('cwuOpportunityAttachments')
+      .where({ opportunityVersion: result.id })
+      .select('file')).map(row => row.id);
+
+    result.addenda = (await connection<{ id: Id }>('cwuOpportunityAddenda')
+      .where({ opportunity: id })
+      .select('id')).map(row => row.id);
+
+    // Remove createdBy/updatedBy for non-admin or non-author
+    if (session?.user?.type !== UserType.Admin &&
+      session?.user?.id !== result.createdBy &&
+      session?.user?.id !== result.updatedBy) {
+        delete result.createdBy;
+        delete result.updatedBy;
+    }
```

Comment:
Can we make this code more DRY?

Reply:
Maybe move it to `rawCWUOpportunitytoCWUOpportunity`

Reply:
I extracted this out into a function.  Couldn't move directly into `rawCWUOpportunitytoCWUOpportunity` since it's also needed in `rawCWUOpportunitySlimtoCWUOpportunitySlim` but I think this is still an improvement.  Let me know if you'd prefer a different approach.